kv_test.go

```diff
@@ -195,6 +197,195 @@ func TestCompareAndDelete(t *testing.T) {
 	assert.True(t, deleted)
 }
 
+func TestSetAtomicWithRetries(t *testing.T) {
+	tests := []struct {
+		name      string
+		key       string
+		valueFunc func(oldValue []byte) (newValue interface{}, err error)
+		setupAPI  func(api *plugintest.API)
+		wantErr   bool
+	}{
+		{
+			name: "Test SetAtomicWithRetries success after first attempt",
+			key:  "testNum",
+			valueFunc: func(old []byte) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(true, nil)
+			},
+		},
+		{
+			name: "Test success after first attempt, old is struct and as expected",
+			key:  "testNum2",
+			valueFunc: func(old []byte) (interface{}, error) {
+				type toStore struct {
+					Value int
+				}
+				var fromDB toStore
+				if err := json.Unmarshal(old, &fromDB); err != nil {
+					return nil, err
+				}
+				if fromDB.Value != 1 {
+					return nil, errors.New("old not as expected")
```

Comment:
would it be better to assert here than returning an error? As i'm afraid that this is relying on the implementation of the method to correctly report this through the API, where in reality it is conceivable that this may not be reported back correctly. Thoughts?

Reply:
> As i'm afraid that this is relying on the implementation of the method to correctly report this through the API, where in reality it is conceivable that this may not be reported back correctly.

I'm not exactly sure what you mean, can you clarify?

Reply:
So here we're returning an error if `fromDB.Value` does not match the value we're expecting, and through that we're asserting that this function should receive the expected `old` value. 

The concern here is that it assumes that this error gets returned through the `SetAtomicWithRetries` API as per its implementation. What if that assumptions was not correct, i.e. what if the implementation is updated to not return that error but rather ignore it, it would then proceed to use that `nil` value in `KVSetWithOptions` and if certain conditions were just right (in the mock setup, etc) this test would pass even though the `oldValue` being passed into this func would not be expected.

My suggestion was to assert directly in this function as opposed to doing it via returning an error. The added benefit of asserting directly in this func is that when it fails you know exactly which condition caused the test to fail, making it easy to debug, rather than figuring out why a certain mock condition wasn't satisfied.

Reply:
First, I don't think it's possible to assert within the call to valueFunc (in a way that would fail the tests). If there is, let me know and I'll do it. It will be quick.

Otherwise, I don't agree that this is a case we need to test for. It's not testing the function, it's testing against every future change that could happen to the code--that's a standard of testing that is never applied to any other PR, so why here? If someone were to change the body of SetAtomicWithRetries so that it ignored the error returned by valueFunc, then it breaks the contract of the function (specified in the comments to the function, and as tested in the tests -- if an error is thrown, it will stop the function and be returned).

Look, I think we're way past the point of high impact here.

Reply:
You can use a function closure to assert, something like:
```
func getValueFunc(t *testing.T) func(old []byte) (interface{}, error) {
	return func(old []byte) (interface{}, error) {
		...
        require.Equal(t, 2, old)
	}
}
```

> It's not testing the function, it's testing against every future change that could happen to the code

Not advocating to test against every future change, that would be impossible :P

Just advocating to lock down the current expected behaviour of the API by asserting using the testing API as opposed to relying on the implementation of the API itself. Especially since this is a plugin API which will be used by many, and therefore has a high impact.

2/5 that we should assert if it's straight forward to do.

Reply:
Okay, done.